go/service/main.go

```diff
@@ -470,6 +471,31 @@ func (d *Service) runBackgroundIdentifierWithUID(u keybase1.UID) {
 	d.G().AddUserChangedHandler(newBgi)
 }
 
+func (d *Service) runBackgroundPerUserKeyUpgrade() {
+	if !d.G().Env.GetUpgradePerUserKey() {
+		d.G().Log.Debug("PerUserKeyBackground disabled (not starting)")
+		return
+	}
+
+	arg := &engine.PerUserKeyBackgroundArgs{
+		Settings: engine.PerUserKeyBackgroundDefaultSettings,
+	}
+	eng := engine.NewPerUserKeyBackground(d.G(), arg)
+	go func() {
+		ectx := &engine.Context{NetContext: context.Background()}
+		err := engine.RunEngine(eng, ectx)
+		if err != nil {
+			d.G().Log.Warning("per-user-key background upgrade error: %v", err)
+		}
+	}()
+
+	d.G().PushShutdownHook(func() error {
+		d.G().Log.Debug("stopping per-user-key background upgrade")
+		eng.Shutdown()
+		return nil
+	})
```

Comment:
Out of curiosity, what happens if we don't shut down the loop? Is the assumption here that we need to cancel background loops or else the process won't be able to exit? I don't _think_ that's true, and I'm going to test it now...